### ISSUE TITLE
array.blockwise has incorrect output shape in task graph

### ISSUE THREAD

**Issue by kortemaki** [dask/dask#6312]
**What happened**: The new_axes parameter only modifies the output array shape at compute time.  Because of this, it's not possible to build task graphs that depend on array.blockwise reshaping the array.

**What you expected to happen**: The output array should be registered as the correct shape in the computation graph, so that the in-dask statements in the example below run the same as the numpy ones.

**Minimal Complete Verifiable Example**:

```python
import dask.array as da
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer().fit_transform([
    "Hello world",
    "Hello bugs",
    "Goodbye world",
    "Goodbye bugs"
])
"""
try to perform the following as a blocked computation in dask:
mat = cosine_similarity(tfidf, tfidf)
mat.shape 
###(4,4)
mat.ravel().shape
### (16,)
mat.ravel()
### array([1. , 0.5, 0.5, 0. , 0.5, 1. , 0. , 0.5, 0.5, 0. , 1. , 0.5, 0. , 0.5, 0.5, 1. ])
"""

"""perform the blockwise function in dask"""
inputs = da.from_array([tfidf[:2,:], tfidf[2:,:]], chunks=(1))
mat = da.blockwise(
    (lambda x,y: cosine_similarity(x[0], y[0])), 'ij',
    inputs, 'i',
    inputs, 'j',
    dtype='f8',
    new_axes={0:4,1:4}
)

print(mat.compute().shape) 
### (4,4)

print(mat.compute().ravel().shape
### (16,)

mat.compute().ravel()
### array([1. , 0.5, 0.5, 0. , 0.5, 1. , 0. , 0.5, 0.5, 0. , 1. , 0.5, 0. , 0.5, 0.5, 1. ])

"""So far so good, now perform the ravel in dask"""

print(mat.shape) 
### (2, 2)

print(mat.ravel().shape)
### (4,)

mat.ravel().compute()
### ValueError: cannot reshape array of size 8 into shape (2,)
```

**Anything else we need to know?**: 

**Environment**: 

- Dask version: 2.18.0
- Python version: 3.6.9
- Operating System: MacOS Catalina 10.15.4
- Install method (conda, pip, source): pip
- sklearn version: 0.20.3

**Comment by TomAugspurger** [comment]
Are your input arrays correct? Do you really want a 1D Dask Array of 2D sparse matricies, or do you rather want a 2D Dask Array?


```python
In [90]: inputs[0]
Out[90]: dask.array<getitem, shape=(), dtype=int64, chunksize=(), chunktype=numpy.ndarray>

In [91]: inputs[0].compute()
Out[91]:
<2x4 sparse matrix of type '<class 'numpy.float64'>'
        with 4 stored elements in Compressed Sparse Row format>
```

If you want a 2D Dask Array, you might want to investigate the `sparse` package. Something like

```python
In [94]: x = sparse.COO.from_scipy_sparse(tfidf[:2, :])

In [95]: a = da.from_delayed(dask.delayed(x), shape=x.shape, meta=x)

In [96]: b = da.from_delayed(dask.delayed(x), shape=x.shape, meta=x)

In [97]: da.concatenate([a, b])
Out[97]: dask.array<concatenate, shape=(4, 4), dtype=float64, chunksize=(2, 4), chunktype=sparse.COO>
```

I don't have time to look further right now unfortunately. But hopefully that gets you started.

As an aside, dask-ml implements FeatureHasher & HashingVectorizer: https://ml.dask.org/modules/api.html#dask-ml-feature-extraction-text-feature-extraction

**Comment by kortemaki** [comment]
Thanks for the response, I'll see if I can use sparse to get past this on the user side and will update once I find something that works.

I'm sure it's possible to compute tfidf in dask by vectorizing the texts and normalizing from there, but it's beyond the scope of the issue here.  As an aside, our input dataset is of manageable size N, it's computations over the N^2 similarity matrix that absolutely need to be distributed.

I still believe it should be possible to define a blockwise operation that changes the number of elements in each chunk (as a result of the computation, not re-chunking).  I have no idea why the new_axes parameter doesn't seem to affect the output shape until the computation is run, and I think it's a bug that you can't chain operations like ravel onto the resulting matrix.

**Comment by kortemaki** [comment]
Alright, the approach suggested by TomAugspurger works as long as you convert the sparse blocks back to scipy representation before calling cosine_similarity!  I'll keep following to see what ultimately happens with this but I'm no longer blocked.  The moral here I guess is if you can at all get your arrays into a native dask format, do so.  I wasn't aware of the sparse.COO chunktype!

```python
mat = da.blockwise(
    (lambda x,y: cosine_similarity(x[0].to_scipy_sparse(), y[0].to_scipy_sparse())), 'ij',
    inputs, 'ik',
    inputs, 'jk',
    dtype='f8',
)
```

**Comment by TomAugspurger** [comment]
> I'm sure it's possible to compute tfidf in dask by vectorizing the texts and normalizing from there, but it's beyond the scope of the issue here. 

Agreed. We can chat about that in https://github.com/dask/dask-ml/issues/5 if you have questions / code it share :)

Glad to hear you got something working. One other important lesson though, is that operations like `.ravel()` might not work if the actual shape of the chunks doesn't match Dask's expectation.

So things like the following:

```python
In [2]: mat
Out[2]: dask.array<lambda, shape=(2, 2), dtype=float64, chunksize=(1, 1), chunktype=numpy.ndarray>

In [3]: mat.shape
Out[3]: (2, 2)

In [4]: mat.compute().shape
Out[4]: (4, 4)
```

are a sign that something has gone wrong. There might still be issues in `blockwise`, but we should addess issues with the block shapes first.
